Modernise setup by moving from setup.py to pyproject.toml

observingClouds opened the pull request:
### Description
This PR modernizes the setup of cfgrib via pyproject.toml, so that one can also use modern tools like `uv` during e.g. development.

This includes the following changes:
- switch from `setuptools` to `hatch` for [good reasons](https://hatch.pypa.io/latest/why/)
- switch from `tox` to `hatch` to simplify setup
- update to current python versions
   - update python classifiers in metadata
   - update test environment
- run tests against wheel and sdist to ensure complete packaging
- fix install dependencies (cfgrib incompatible with latest xarray)
- combine all dependency group in `pyproject.toml`
- remove (outdated) frozen environments
- remove weekly environment update action (has been broken for over a year)
- add weekly functionality test with latest dependencies

Please excuse this rather large PR. The setup was quite antique (e.g. using `xarray==0.15.1` and python 3.7), so quite a few changes were necessary. Nevertheless, tests should cover all the changes and improved the overall stability of this project.

### Contributor Declaration

By opening this pull request, I affirm the following:

* The code follows the project's coding standards.
* I have performed self-review and added comments where needed.
* I have added or updated tests to verify that my changes are effective and functional.
* I have run all existing tests and confirmed they pass.

observingClouds committed:
move setup completely to pyproject.toml

observingClouds committed:
adjust formatting

observingClouds committed:
set README

observingClouds committed:
add black config

observingClouds committed:
document changes

DWesl commented on pyproject.toml:
```diff
@@ -1,5 +1,66 @@
+path = "cfgrib/__init__.py"
+
+[tool.hatch.build.targets.sdist]
+include = ["/cf2cdm"]
```
```suggestion
[tool.hatch.build.targets.wheel]
include = ["cf2cdm", "cfgrib"]
```

I ran into problems attempting to build this branch, and the suggestion hatch printed looked a lot like the change above

observingClouds replied:
I have added both `cf2cdm` and `cfgrib` to `sdist` and `wheel` targets now.

DWesl commented on setup.py:
```diff
@@ -1,74 +0,0 @@
-    license="Apache License Version 2.0",
-    url="https://github.com/ecmwf/cfgrib",
-    packages=setuptools.find_packages(),
-    include_package_data=True,
```
This setting doesn't get carried over and could, but on the other hand, I don't think there's actually any package data files *to* include, so dropping this might be the right thing to do.

observingClouds replied:
I implemented a packaging tests that runs the test suite against the wheel and sdist to ensure that nothing is missing from the package. Now unnecessary files, like tests, grib and netCDF files are no longer included in the wheel, dropping the size of the package from around 9MB to less than 50KB 🥳. So, yes `include_package_data` is not necessary as the tests show.

DWesl commented on setup.py:
```diff
@@ -1,74 +0,0 @@
-    version=parse_version_from("cfgrib/__init__.py"),
-    description="Python interface to map GRIB files to the NetCDF Common Data Model "
-    "following the CF Convention using ecCodes.",
-    long_description=read("README.rst") + read("CHANGELOG.rst"),
```
README.rst made it over, CHANGELOG.rst did not.  Setuptools seems to have [a method](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#dynamic-metadata) for doing this, but [hatch explicitly disallows that approach](https://github.com/pypa/hatch/blob/c7575ad551f380bd2a1078bb5ba2855bf961a659/backend/src/hatchling/metadata/core.py#L560-L563)

observingClouds replied:
Yes, according to [PEP-621](https://peps.python.org/pep-0621/#readme) `pyproject.toml`'s `readme` key is the replacement for `long_description` of setuptools, but does not support the merging of text at its assignment.

However, https://github.com/hynek/hatch-fancy-pypi-readme is build to do exactly this. I'll adjust the PR.

observingClouds replied:
Has been addressed in e76753920f82bdcb7e8b642a4856b4da810fe8b7

DWesl commented on pyproject.toml:
```diff
@@ -1,5 +1,66 @@
 [build-system]
-requires = [ "setuptools>=61", "setuptools-scm>=8" ]
```
There is [a hatch VCS version plugin](https://github.com/ofek/hatch-vcs#readme), though the old `setup.py` specified `version`, so I'm not sure setuptools_scm's versioner actually got used: I certainly don't see the configuration required to get the version in the sdist and wheel filenames with the old version.

This might be why `.pre-commit-config.yaml` got dropped from the sdist, which is the big difference between the created files (Setuptools-scm sorts alphabetically and includes entries for each directory, while hatch sorts files before directories, except for a list of five specific files, but that should be a minor difference)

EDIT: I just realized: if you split off the `hatchling` change and try to keep the [current version determination](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#dynamic-metadata), I'm not sure if the setuptools version or the setuptools_scm version will wind up getting used.  (EDIT: If a version is specified in the project table, even dynamically, that one is used.  If there are additionally directions on how setuptools_scm should get a version, it warns that those directions won't be used).

observingClouds replied:
The version seems to have always been extracted from `cfgrib/__init__.py`. The new format just does not require the [boilerplate code](https://github.com/ecmwf/cfgrib/blob/master/setup.py#L30-L35), as this is integrated into [hatch](https://hatch.pypa.io/1.12/version/#pyprojecttoml)

DWesl reviewed (commented):
Looks good overall, although I'm curious about the combined shift from `setup.py` to `pyproject.toml` and from `setuptools` to `hatchling`.

Are there any benefits to `hatchling` over `setuptools` beyond hatchling's integration with `hatch`?

observingClouds committed:
Merge README and CHANGELOG to one repo description

observingClouds committed:
Merge remote-tracking branch 'origin/master' into pr/438

observingClouds committed:
add cfgrib to build

observingClouds committed:
move test dependencies to correct group

observingClouds committed:
run pytest against packages

observingClouds committed:
Merge branch 'maint/pyproject' of github.com:observingClouds/cfgrib into maint/pyproject

observingClouds committed:
update python classifiers

observingClouds committed:
update tox

observingClouds committed:
get necessary tox.ini

observingClouds committed:
ignore wheel having multple toplevel entries

observingClouds committed:
get tests

observingClouds committed:
fix casting error of datetime64/timedelta64

datetime64/timedelta64 need to be casted to int32 as netCDF3 (used by scipy) does not support int64

observingClouds committed:
allow for floating point differences

observingClouds committed:
fix tests by restricting xarray

observingClouds committed:
consolidate environments

observingClouds committed:
change weekly test to run actual test with latest environment

observingClouds committed:
fix weekly test syntax

observingClouds committed:
convert tox env to hatch in pyproject

observingClouds committed:
use core python package

observingClouds committed:
add build docs gh action

observingClouds committed:
apply precommit

observingClouds committed:
Merge remote-tracking branch 'fork/maint/pyproject' into maint/pyproject

observingClouds committed:
move coverage report to extra job; add precommit autofix

observingClouds committed:
revert adding precommit action

DWesl commented on .github/workflows/weekly.yml:
```diff
@@ -7,38 +7,17 @@ on:
       with:
-        auto-update-conda: true
-        python-version: ${{ matrix.python }}
-        activate-environment: ${{ matrix.os }}-${{ matrix.python }}${{ matrix.extras }}
-        environment-file: environment${{ matrix.extras }}.in.yml
-    - name: Export concrete dependencies
-      shell: bash -l {0}
-      run: |
-        conda env export --no-build -f tests/environment-${{ matrix.os }}-${{ matrix.python }}${{ matrix.extras }}.yml
-        git diff
```
This, I think, would be a `uv compile` step

DWesl commented on .github/workflows/weekly.yml:
```diff
@@ -7,38 +7,17 @@ on:
     steps:
     - uses: actions/checkout@v2
-    - uses: conda-incubator/setup-miniconda@v2
+    - uses: astral-sh/setup-uv@v5
```
Have you confirmed that uv can handle installing the ecCodes C library in addition to the ecCodes python package?

DWesl commented on CHANGELOG.rst:
```diff
@@ -5,20 +5,23 @@ Changelog for cfgrib
 unreleased
 ----------
 
+- Deprecating setuptools in favor of hatchling and pyproject.toml.
```
Nothing you do here will deprecate setuptools, which is still a valid build backend (unlike distutils, which was deprecated some years ago and has since been removed from Python).  Among other things, setuptools is one of three backends that supports compiled extension modules.  Hatchling, so far, only seems to support Cython extension modules.

I would then suggest:
```suggestion
- Move away from discouraged `setup.py` in favor of `pyproject.toml`, 
  and also replace setuptools with hatchling as the build backend
```
and then this PR does more things, which might need to be documented in the changelog:
- overhaul CI test setup to drop Appveyor in favor of GitHub Actions
- set default encoding for time variables
- drop support and testing for Python<3.10
- drop tox configuration

DWesl commented on CHANGELOG.rst:
```diff
@@ -5,20 +5,23 @@ Changelog for cfgrib
 - Set default dtype for step to `timedelta64[ns]` for compatibilty with
   future versions of xarray.
   See `#427 <https://github.com/ecmwf/cfgrib/pull/427>`_.
 
-- Fixed  xarray.backends imports for compatibilty with recent versions of xarray.
-  See `#409 <https://github.com/ecmwf/cfgrib/pull/409>`_.
-
```
Is there a reason you changed the order?

DWesl commented on tests/test_50_xarray_getitem.py:
```diff
@@ -39,7 +40,7 @@ def test_getitem_slice(cache: bool) -> None:
     # NOTE: label based indexing in xarray is inclusive of both the start and stop bounds.
     assert da.sel(number=slice(2, 6)).mean() == va[2:7].mean()
-    assert da.sel(number=slice(2, 6, 2)).mean() == va[2:7:2].mean()
+    assert np.isclose(da.sel(number=slice(2, 6, 2)).mean(), va[2:7:2].mean())
```
Could you elaborate on why this was changed?  Those two expressions should be performing exactly the same operations: averaging the third, fifth, and seventh elements of the array used to form the DataArray; there is an excellent chance XArray forwards to the NumPy method used on the right.

In addition, pytest has an `approx` function to do what you use `np.isclose` for here (that is, `da.sel(...).mean() == approx(va[...].mean(), ...)

DWesl commented on tox.ini:
```diff
@@ -1,22 +0,0 @@
-[tox]
-envlist = qc, docs, py37, pypy3, deps
-
-[testenv]
-passenv = WHEELHOUSE PIP_FIND_LINKS PIP_WHEEL_DIR PIP_INDEX_URL
-setenv = PYTHONPATH = {toxinidir}
-deps = -r{toxinidir}/ci/requirements-tests.txt
-commands = pytest -v --flakes --cache-clear --basetemp={envtmpdir} {posargs}
-
-[testenv:docs]
-deps = -r{toxinidir}/ci/requirements-docs.txt
-commands = sphinx-build -W -b html docs build/sphinx/html
-
-[testenv:qc]
-basepython = python3.7
-# needed for pytest-cov
-usedevelop = true
-commands = pytest -v --flakes --pep8 --mccabe --cov=cfgrib --cov=cf2cdm --doctest-glob="*.rst" --cov-report=html --cache-clear --basetemp={envtmpdir} {posargs}
-
-[testenv:deps]
-deps =
-commands = python setup.py test
```
Could you elaborate on why you got rid of the tox configuration?  The dependencies would look a little different, but the ability to run all tests and linters with a single command is useful, especially for beginning contributors who don't know what to look for.

DWesl reviewed (commented):
I like the move from `setup.py` to `pyproject.toml`, and think that part of this PR is in a good place.

I kind of understand the move from setuptools to hatchling, though I would have left that for a second PR.  I think that part also works.

Dropping the `requirements*.txt` files listing the versions of each Python package used in testing is a choice.  The primary relation to the rest of the PR is that the `requirements*.in` files got moved to sections of `pyproject.toml`.

There's a collection of CI changes that would probably work better as a separate PR, to show that the tests pass with the current setup and also that this PR doesn't break those tests.

Python 3.9 and all those before it have reached end of support, so it's probably time to drop support for those versions, but I don't know that a PR about moving from `setup.py` to `pyproject.toml` and `setuptools` to `hatchling` is the place to do it.

I don't see a replacement for the tox functionality that was removed: does uv have something similar?  Is there a way to run GitHub Actions from the command line?

DWesl commented on cfgrib/__main__.py:
```diff
@@ -140,6 +140,24 @@ def to_netcdf(
+            if "datetime64" in dtype_str:
+                netcdf_kwargs["encoding"][var_name] = {
+                    "units": "seconds since 1970-01-01",
+                    "dtype": "int32",
```
Out of curiosity, do you know what the GRIB format uses for these variables?

DWesl commented on .github/workflows/packaging-test.yml:
```diff
@@ -0,0 +1,73 @@
+        uses: actions/checkout@v6
+        with:
+          sparse-checkout: |
+            tox.ini
```
This file got deleted

DWesl reviewed (commented):
The CI overview informs me that the CLA check is failing: have you completed and submitted the Contributor License Agreement?  (Can one of the maintainers point to where the docs show where that is?)

You would also need to edit `CONTRIBUTING.rst` to reflect that `tox` will no longer run tests in a collection of Python environments and also run the linters and build the documentation.  Editing the `Makefile` might work, but I'm not sure how many Python developers have `make`.
